content/blog/2023/2023-08-31-a-dash-of-dask.md

```diff
@@ -0,0 +1,45 @@
+
+Ideally, this kind of data handling should be separate from the canvas and widgets. Both to facilitate development and to not obtrude the end user. Also, we need the learners to maintain backward compatibility because Orange is quite a large project. These requirements mean some _magic_ has to be done behind the scenes to account for the subtle differences between scikit-learn and dask-ml. So, to anyone trying to use one of the learners mentioned above in unpredicted ways, you have been warned.
+
+Additionally, some decisions had to be made about when operations should be computed and what should be stored in memory. Consider, for example, this scenario:
+
+You have made some changes to logistic regression, so you run some tests to ensure everything works, but it takes a while to finish. Waiting is not surprising since you are processing a large amount of data, which usually takes time. But sometimes you wonder if you might be waiting a little too long. So you think, 'Maybe I've made a mistake'. You decide to go through every phase of the computation painstakingly and ensure nothing takes an unreasonable amount of time. After a while, you happen upon a preprocessor that, at first glance, seems to be doing little computing but is taking inordinately long to finish! The preprocessor is called RemoveNaNColumns, and its job is to eliminate columns with too many missing values like so:
+
+```python
+nans = np.sum(np.isnan(data.X), axis=0)
+att = [a for a, n in zip(data.domain.attributes, nans) if n < threshold]
+```
+
+This specific bit of code hasn't been touched up to work with Dask, but it still works (kind of). Typically, it would sum up the number of unknown values along the 0-th axis, then loop through them, removing attributes with too many. The catch with Dask is that the first part of this process is only computed when it must be. Since the first part is vectorized and the second isn't, the entire calculation (for all columns) is performed repeatedly in each iteration, slowing the process down substantially.
+
+The ideal solution would be to vectorize the second part, but sadly, that is not something I know how to do with Python lists. So you are left with option one, doing the first part in a loop, or option two, computing the first part ahead of time and storing the results in memory. Here's an example of how each of these approaches might look.
+
+```python
+# option 1
+att = [a for i, a in enumerate(data.domain.attributes)
+       if np.sum(np.isnan(data.X[:, i])) < threshold]
+
+# option 2
+nans = np.asarray(np.sum(np.isnan(data.X), axis=0))
+att = [a for a, n in zip(data.domain.attributes, nans) if n < threshold]
+```
+
+In theory, option one should have a smaller memory impact, and some version of it may be preferred under certain circumstances. However, we've decided to go with option two for our purposes.
```

Comment:
Here we need to explain why that the sum of everything together is faster with dask.